[1단계 - 장기] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
안녕하세요 이스트!
백엔드 7기 듀이입니다.
장기 미션에서 만나뵙게 되어 반갑습니다!
리뷰 잘 부탁드립니다~!😄

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/4083

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- ``BoardGenerator``에서 보드를 생성하는 과정과 ``Ma``, ``Sang``의 이동 경로를 구현하는 부분에서 하드코딩이 많고 이중 for문이 생겨 indent를 지키거나 메서드 라인 제한을 지키지는 못했습니다. (뷰에서도 indent를 지키지는 못했습니다..😭)
- 다만 이 원칙들은 객체지향을 위해 최소한의 장치를 걸어둔 것일뿐, 설계를 할 때나 코드를 작성할 때 최대한 객체지향을 지키려고 노력했고, 생성하는 객체에게 책임을 잘 부여했는가 생각하면서 미션을 진행했기 때문에 3점 정도 충족했다고 생각합니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 추상 클래스와 인터페이스
- 장기 말들의 구현체는 인터페이스인 ``Piece``를 구현하고 있습니다.
처음엔 페어와 함께 ``Piece``를 추상 클래스로 선언했지만, 아래와 같은 이유들로 인터페이스로 변경했습니다.
  - 인터페이스의 장점 : 객체의 행동만 추상화 할 수 있다. 개방 폐쇄 원칙을 지킬 수 있다. 다중 상속이 가능하다.
  - 인터페이스의 단점 : 자식 클래스에서 중복 코드가 생길 수 있다.
  - 추상 클래스의 장점 : 부모를 상속하는 자식들에게 공통된 상태와 행동을 강제할 수 있어 중복 코드를 줄일 수 있다.
  - 추상 클래스의 단점 : 자식 클래스가 부모 클래스에 강결합되어 있어 변경에 취약하다. 다중 상속이 불가하다.
  - 인터페이스의 장점이 추상 클래스의 단점을 감수할 만 하다고 생각하여 인터페이스를 선택했습니다.

### ``FlowManager``와 ``JanggiManager``
- ``controller``를 사용하지 않고, 도메인의 책임에 집중하기 위해 ``FlowManager``와 ``JanggiManager``를 두었습니다.
- 각 객체의 책임은 아래와 같습니다.
  - ``FlowManager``: 뷰와 상호작용하며, 프로그램의 흐름을 관리합니다. 흐름에 따라 도메인 로직과 관련된 부분은 ``JanggiManager``에게 요청합니다.
  - ``JanggiManager``: 장기 게임에 필요한 ``Board``와 상호작용하며 게임과 관련된 주요 도메인 로직을 실행합니다.

### 말의 이동 규칙
- ``Piece``인터페이스를 상속하는 말들은 ``canMove``로 움직일 수 있는지만 강제하고, 이동 규칙은 각 말에서 구현하도록 설계했습니다.
한 단계 더 나눈다고 하면 같은 움직임의 패턴을 가진 ``Ma``, ``Sang``을 ``DiagonalPiece``로, 그 외의 기물들을 ``StraightPiece`` 정도로 추상화할 수 있을 것 같긴 한데요.
이번 단계에서는 굳이 필요하지 않은 부분이라고 생각하여 ``Piece``까지만 추상화했습니다.
좀 더 추상화를 하는 것이 좋을지 그대로 두어도 괜찮을지 이스트의 의견은 어떠실지 궁금합니다.

소중한 시간 들여 코드와 질문 읽고 리뷰해주셔서 감사드립니다!😊👍🏻

ljhee92 committed:
docs: 프로그램 실행 흐름 기능 작성

threepebbles committed:
feat(Point): 좌표 클래스 추가

threepebbles committed:
feat(Direction): 방향을 의미하는 상수 추가

threepebbles committed:
feat(Node, Edge): 노드, 엣지 클래스 추가

threepebbles committed:
feat(BoardInitializer): 노드와 엣지 정보 초기화

ljhee92 committed:
feat(SangMaOrderCommand): 상마 순서를 의미하는 enum 추가

ljhee92 committed:
feat: 팀을 구분하기 위한 enum 추가

ljhee92 committed:
feat(Piece): 기물 역할 클래스 추가

ljhee92 committed:
feat(BoardInitializer): 초나라 기물 배치 초기화 기능 추가

ljhee92 committed:
feat(BoardInitializer): 상마 순서에 따른 기물 배치 초기화 기능 추가

threepebbles committed:
refactor: BoardInitializer 리팩터링

threepebbles committed:
refactor: 패키지 이동

threepebbles committed:
feat: 특정 좌표에 기물 존재 여부 확인 기능 추가

threepebbles committed:
feat(Piece): 각 기물의 종류를 구분하는 기능 추가

threepebbles committed:
feat(OutputView): 장기판 출력 기능 추가

ljhee92 committed:
feat(OutputView): 팀 색깔 구분 기능 추가

ljhee92 committed:
refactor: 기물이 가진 팀을 묻도록 리팩터링

threepebbles committed:
feat(Wang): 왕의 이동 가능 여부 확인 기능 추가

ljhee92 committed:
feat(Sa): 사의 이동 가능 여부 확인 기능 추가

ljhee92 committed:
fix: 노드 엣지 초기화 기능 수정

ljhee92 committed:
feat(Byeong): 병의 이동 가능 여부 확인 기능 추가

threepebbles committed:
feat(Cha): 차의 이동 가능 여부 확인 기능 추가

threepebbles committed:
feat(Ma): 마의 이동 가능 여부 확인 기능 추가

ljhee92 committed:
feat(Sang): 상의 이동 가능 여부 확인 기능 추가

threepebbles committed:
refactor: 변수명 개선, 인덴트 깊이 개선

threepebbles committed:
refactor: 가독성 개선

threepebbles committed:
fix: 테스트 코드 컴파일 에러 수정

threepebbles committed:
feat(Po): 포의 이동 가능 여부 확인 기능 추가

ljhee92 committed:
feat: 상마 순서 입력 기능 추가

ljhee92 committed:
feat: 이동할 기물의 위치 입력 기능 추가

ljhee92 committed:
refactor(Board): 중복 메서드 제거, 네이밍 개선

ljhee92 committed:
feat(OutputView): 승패 결과 출력 기능 추가

ljhee92 committed:
feat(JanggiManager): 말을 움직이는 기능 추가

ljhee92 committed:
feat: 순서대로 장기 게임을 진행하는 기능 추가

threepebbles committed:
feat(ErrorHandler): 메서드 실행 중 예외 발생 시 재실행하는 기능 추가

threepebbles committed:
refactor(Board): 특정 위치의 놓인 기물의 종류를 확인하는 기능 추가

threepebbles committed:
test: Board 테스트 추가

threepebbles committed:
fix: 특정 위치의 기물의 종류를 파악하는 기능 수정

threepebbles committed:
fix: 턴에 맞는 기물을 움직일 수 있게 수정

threepebbles committed:
refactor: 용어 수정(노드 -> 위치)

threepebbles committed:
test: Node 테스트 추가

threepebbles committed:
refactor: 네이밍 개선, 메서드 분리

threepebbles committed:
refactor: 패키지 이동

threepebbles committed:
refactor: 네이밍 개선, 네이밍 컨벤션에 맞게 수정

threepebbles committed:
refactor(Turn): 내부 static 클래스로 전환

threepebbles committed:
refactor: 네이밍 개선

threepebbles committed:
refactor(Node): edges() 제거

threepebbles committed:
refactor: 메서드 내 변수 상수화

threepebbles committed:
refactor: 변수명 변경

ljhee92 committed:
refactor: 상수 final 처리, ERROR prefix 추가

threepebbles committed:
Merge branch 'step1' of https://github.com/ljhee92/java-janggi into step1

EastHShin commented on src/main/java/domain/board/BoardGenerator.java:
```diff
@@ -0,0 +1,156 @@
+    private boolean isInRange(int row, int column) {
+        return START_ROW_INDEX <= row && row <= END_ROW_INDEX
+                && START_COLUMN_INDEX <= column && column <= END_COLUMN_INDEX;
+    }
```
Point 에서의 범위 검증 로직과 중복으로 존재하네요 :)
row, col 에 대한 범위 검증은 Point 에 모두 위임하고, BoardGenerator 에서는 검증된 Point 를 사용하는 것은 어떨까요?

ljhee92 replied:
오 그렇네요!
이스트의 질문을 받고 ``Board`` 내부에서 검증할까 생각도 해보았는데요,
항상 검증된 ``Point``를 사용하는 것이 더 안전하다고 판단되어 ``Point`` 내부에서 검증할 수 있도록 수정했습니다.

[refactor: Point의 범위 검증은 Point 내부에서 수행하도록 변경](https://github.com/ljhee92/java-janggi/commit/dfb0e1a46d6426506d0310a20d3a76757ea68117)

EastHShin commented on README.md:
```diff
@@ -1,3 +1,92 @@
+### 이동 커맨드 입력
+
+- move src dst 양식으로 입력해야 한다.
+- src에 있는 내 기물을 dst로 이동한다.
```
<img width="265" alt="image" src="https://github.com/user-attachments/assets/b1f77f1f-368c-4ece-8b35-f9e48bce12f1" />

동일한 칸으로 이동할 경우 해당 기물이 스스로 잡아먹히게 되네요ㅎㅎ

ljhee92 replied:
하핫,, 이 부분은 기능을 만들지 않았었네요!
추가해두었습니다!

[feat: 같은 위치로 이동할 수 없는 기능 추가](https://github.com/ljhee92/java-janggi/commit/b44459125a6e0ae37872c0d298d5494060fe865d)

EastHShin commented on src/main/java/domain/board/Node.java:
```diff
@@ -0,0 +1,73 @@
+
+    public boolean hasEdgeByDirection(Direction direction) {
+        return edges.stream()
+                .anyMatch(edge -> edge.direction() == direction);
```
Edge 도메인에 isSameDirection 이라는 메세지를 보내는건 어떨까요?

ljhee92 replied:
direction을 꺼내와서 확인하는 것보다 ``Edge`` 도메인에게 넘겨주고 확인하는 방식이 훨씬 객체지향적이네요!
감사합니다😄

[refactor: 같은 방향인지 확인하는 책임을 Edge 객체에게 위임](https://github.com/ljhee92/java-janggi/commit/9a78f3dd58fade1e54fabbea1b1c7d4843b2f5ab)

EastHShin commented on src/main/java/FlowManager.java:
```diff
@@ -0,0 +1,72 @@
+import view.OutputView;
+import view.SangMaOrderCommand;
+
+public class FlowManager {
```
>controller를 사용하지 않고, 도메인의 책임에 집중하기 위해 FlowManager와 JanggiManager를 두었습니다.
각 객체의 책임은 아래와 같습니다. FlowManager: 뷰와 상호작용하며, 프로그램의 흐름을 관리합니다. 흐름에 따라 도메인 로직과 관련된 부분은 JanggiManager에게 요청합니다. JanggiManager: 장기 게임에 필요한 Board와 상호작용하며 게임과 관련된 주요 도메인 로직을 실행합니다.

듀이는 컨트롤러의 역할이 무엇이라고 생각하시나요? 저는 뷰와 도메인을 사용하며 어플리케이션의 흐름을 제어하는 역할이라고 생각합니다. 컨트롤러를 없애고 FlowManager 를 만들어주셨지만, FlowManager 가 하는 일이 곧 컨트롤러의 역할인 것 같아요. MVC 패턴으로 로직을 작성하는 이상, View와 Model 이 존재하기 때문에 이를 제어하는 컨트롤러의 역할은 필수로 있어야한다고 생각해요. 컨트롤러가 필요없다면 도메인에 View 의 로직이 이곳저곳에 존재하고 있는 상황일 것 같네요ㅎㅎ

ljhee92 replied:
이스트의 의견에 동의합니다!
저 역시도 컨트롤러는 뷰와 도메인을 연결해주고, 어플리케이션의 흐름을 제어하는 역할이라고 생각해요.
이번 미션에서 페어와 함께 진행하면서 처음으로 객체에 집중해보고자 ``FlowManager``를 도입했는데, 다시 생각해 보니 이렇게 되면 도메인 내부에 뷰의 로직이 존재하고 있는 게 되겠네요.
domain과 view를 분리한 이상 MVC 패턴을 사용하는 것이고, 다른 사람들도 익숙한 MVC 패턴에 맞춰 controller로 명시해주는 것이 더 확실한 책임 분리가 되겠다고 생각하여 수정해두었습니다.

[refactor: FlowManager -> JanggiController로 수정, Turn 객체 분리](https://github.com/ljhee92/java-janggi/commit/21d43fbf73f80e70601df1f3e1f5be34e976f6d9)

EastHShin commented on src/main/java/FlowManager.java:
```diff
@@ -0,0 +1,72 @@
+        });
+    }
+
+    private static class Turn {
```
위 코멘트에 이어서, Turn이라는 클래스는 도메인 규칙에 밀접해보이는데요, FlowManager라는 컨트롤러에 존재하는 것이 맞을지 생각해보면 좋겠습니다.

ljhee92 replied:
``Turn`` 객체가 도메인 규칙에 더 밀접해보인다는 말씀에 동의합니다.
컨트롤러에 존재하던 내부 클래스를 별도 객체로 분리했습니다.

[refactor: FlowManager -> JanggiController로 수정, Turn 객체 분리](https://github.com/ljhee92/java-janggi/commit/21d43fbf73f80e70601df1f3e1f5be34e976f6d9)

EastHShin commented on src/main/java/domain/piece/Ma.java:
```diff
@@ -0,0 +1,97 @@
+        return findMovableNodes(source, board).contains(destination);
+    }
+
+    private List<Node> findMovableNodes(Node sourceNode, Board board) {
```
>Piece인터페이스를 상속하는 말들은 canMove로 움직일 수 있는지만 강제하고, 이동 규칙은 각 말에서 구현하도록 설계했습니다.
한 단계 더 나눈다고 하면 같은 움직임의 패턴을 가진 Ma, Sang을 DiagonalPiece로, 그 외의 기물들을 StraightPiece 정도로 추상화할 수 있을 것 같긴 한데요.
이번 단계에서는 굳이 필요하지 않은 부분이라고 생각하여 Piece까지만 추상화했습니다.
좀 더 추상화를 하는 것이 좋을지 그대로 두어도 괜찮을지 이스트의 의견은 어떠실지 궁금합니다.

네 말씀주신대로 Ma, Sang에서 완전히 같은 로직이 반복되고 있네요!
이런 경우에는 말씀 주신것처럼 Piece 를 상속받은 DiagonalPiece 를 기반으로 Ma, Sang 으로 만들어주면 어떨까 싶습니다.
인터페이스를 계속해서 사용하실거라면 인터페이스의 default 메서드를 사용하는것도 괜찮을 것 같아요 :)

ljhee92 replied:
우선 이번 단계에서는 ``Piece``의 추상화를 좀 더 확실히 짚고 넘어가야할 것 같아서 아직 ``DiagonalPiece``로 추상화하지는 않았습니다.
대신 default 메서드를 사용해서 기물들 내부에 존재하는 중복 로직을 추상화했습니다.

[refactor: 중복 코드 interface의 default method로 통합](https://github.com/ljhee92/java-janggi/commit/6962760da2d8c698f5665eeae404bad9053355d8)

EastHShin commented on src/main/java/domain/piece/Cha.java:
```diff
@@ -0,0 +1,63 @@
+        return candidates;
+    }
+
+    private void findCandidates(Node currentNode,
```
반복문 구조로는 어떻게 바꿔볼 수 있을까요?

ljhee92 replied:
재귀호출의 경우 스택오버플로우가 발생할 수 있다는 점을 몰랐습니다.
좀 더 안전하게 반복문으로 변경했습니다!

[refactor: 재귀호출 -> 반복문으로 변경](https://github.com/ljhee92/java-janggi/commit/af0d214e90eb5dc31ec58f5dc0ceaa993d4e35a6)

EastHShin commented on src/test/java/domain/board/BoardTest.java:
```diff
@@ -0,0 +1,152 @@
+
+    @Nested
+    @DisplayName("예외가 발생하는 테스트")
+    class Fail {
```
자기자신이 제거되지 않는 테스트가 추가되면 좋겠네요!

ljhee92 replied:
위에 답변 드린 것처럼 기능 구현하면서 테스트 함께 추가해두었습니다!

[feat: 같은 위치로 이동할 수 없는 기능 추가](https://github.com/ljhee92/java-janggi/commit/b44459125a6e0ae37872c0d298d5494060fe865d)

EastHShin reviewed (changes requested):
안녕하세요 듀이~

이번 미션 함께하게 된 이스트입니다!
어려운 미션인데 잘 구현해주셨네요 👍 
코멘트 몇개 남겨두었으니 확인 부탁드려요 :)

ljhee92 committed:
refactor: Point의 범위 검증은 Point 내부에서 수행하도록 변경

- 엄격히 책임을 따지자면 Point의 범위가 Board의 범위에 맞춰 검증되어야 하므로 Board에서 검증하는 것이 맞지 않을까? 생각했습니다.
하지만, 항상 유효한 Point 객체를 사용하고 싶기 때문에 Board 내부에서 검증하지 않고 Point 내부에서 Board 범위에 따라 검증하도록 했습니다.

ljhee92 committed:
chore: 사용하지 않는 .gitkeep 파일 제거

ljhee92 committed:
test: 클래스별 테스트명 명시, 테스트가 표시될 때 언더바(_) 제거되도록 수정

ljhee92 committed:
refactor: 같은 방향인지 확인하는 책임을 Edge 객체에게 위임

ljhee92 committed:
test: Assertions static import로 수정

ljhee92 committed:
feat: 같은 위치로 이동할 수 없는 기능 추가

ljhee92 committed:
refactor: 중복 코드 interface의 default method로 통합

ljhee92 committed:
refactor: FlowManager -> JanggiController로 수정, Turn 객체 분리

- FlowManager의 역할이 뷰와 도메인 사이에서 상호작용하는 controller의 역할과 다르지 않다고 생각하여 JanggiController로 수정했습니다.
- Turn을 도메인으로 판단, 별도 객체로 분리했습니다.
- 내 기물을 이동한 이후에 상대팀의 왕이 죽었는지 확인하도록 수정했습니다.

ljhee92 committed:
refactor: 재귀호출 -> 반복문으로 변경

- 재귀호출은 스택오버플로우의 위험이 있으므로 반복문으로 변경했습니다.

ljhee92 committed:
refactor: 상마 순서 입력 메시지 수정

- 상마 순서 번호를 입력할 수 있도록 메시지를 수정했습니다.

ljhee92 committed:
fix: 차와 포 후보 추가 로직 수정

- 차의 경우 현재 노드를 검증하지 않고 후보에 포함하고 있었으므로 검증 후 추가하도록 수정했습니다.
- 포 역시 현재 노드의 검증이 모두 종료된 후 후보에 추가하도록 수정했습니다.

ljhee92 committed:
refactor: 컨트롤러 로직 수정

- 상마커맨드 입력 시 올바르지 않게 입력했다면 무조건 처음부터 재입력 받던 로직을 각각 재입력 받을 수 있도록 수정했습니다.

- isRunning 변수를 사용하지 않고 do~while을 사용해 왕이 죽었을 때 반복문이 종료되도록 수정했습니다.

ljhee92 committed:
refactor: 매서드 파라미터 final 통일, Board의 메서드 순서 변경

- 메서드 내에서 변하지 않는 파라미터는 final을 붙이도록 통일했습니다.

- Controller의 호출 순서대로 Board의 메서드 순서를 변경했고, 중복되는 메서드(existsPieceTypeByNode, hasPieceTypeByNode)를 하나로 합쳤습니다.

ljhee92 committed:
chore: 오타 수정

ljhee92 committed:
chore: util 패키지 위치 수정

ljhee92 commented:
안녕하세요 이스트,
리뷰해주신 부분 반영해서 푸시해두었습니다.
장기 자체를 이번 미션에서 처음 접하게 되어서 룰도 이해가 어려운데 코드 적용까지 하려니 너무너무 어렵네요..🥹
잘 하고 있는지 모르겠습니다🥲
그래도 열심히 하고 있으니 피드백 반영하는 속도가 조금 더디더라도 양해해주시면 감사드리겠습니다.

중간에 생긴 coderabbitai는 왜 생긴건지 모르겠습니다..! 혹시 알고 계시나요?
아마 우테코 레포지토리에 설정이 되어있는 듯 한데 신경쓰지 않아도 괜찮은 걸까요?🤔

항상 소중한 시간 들여 리뷰해주셔서 감사합니다.
즐거운 한 주 되세요!😄

kang-hyungu commented:
coderabbitai 코멘트는 그냥 무시하시면 됩니다~ coderabbit 테스트 중이에요

ljhee92 committed:
refactor: final 누락된 곳 추가

ljhee92 commented:
다시 보다가 ``final`` 키워드 누락된 곳이 몇 군데 보여서 추가 푸시 한 번 했어요 이스트!
혹시 읽고 계셨다면 죄송합니다🥹

EastHShin commented on src/main/java/domain/piece/Byeong.java:
```diff
@@ -0,0 +1,54 @@
+        return PieceType.BYEONG;
+    }
+
+    @Override
+    public boolean hasTeam(final Team team) {
+        return this.team == team;
+    }
```
모든 구현체에서 같은 동작을 하고 있는데, 요것도 default 메서드로 지정하는건 어떤가요?

ljhee92 replied:
안 그래도 그 생각을 안 한 것은 아니었는데요!
Piece가 추상 클래스가 아니고 인터페이스이다보니 인스턴스 필드를 가지고 있지 않아서 각자 구현하도록 해두었습니다.

ljhee92 replied:
2단계 구현하다 보니 Piece를 추상 클래스로 변경하게 되어 중복 코드 삭제할 수 있었습니다!

[refactor: Piece 인터페이스 -> 추상 클래스로 변경](https://github.com/ljhee92/java-janggi/commit/88f3f17cb4cffbae8d48e016b9f177c25b9c2a01)

EastHShin commented on src/main/java/controller/JanggiController.java:
```diff
@@ -0,0 +1,56 @@
+        Board board = createJanggiBoard();
+
+        Turn turn = new Turn();
+        do {
+            OutputView.printBoard(board);
+            moveByTurn(turn, board);
+            if (board.isOpponentWangDead(turn.team())) {
+                OutputView.printBoard(board);
+                OutputView.printMatchResult(turn.team());
+                break;
+            }
+            turn.changeTurn();
+        } while (true);
```
do-while문일 필요가 있나요?

ljhee92 replied:
오 그렇네요! while문의 조건을 이것저것 변경하면서 테스트하다 보니 마지막에 고려하지 못한 것 같습니다.
반영해두었습니다!

[refactor: do~while문 -> while문 변경](https://github.com/ljhee92/java-janggi/commit/ae0f3c46bc39715796b2487ce8ae9858e84a50a9)

EastHShin commented on src/main/java/domain/board/Node.java:
```diff
@@ -0,0 +1,82 @@
+                .anyMatch(edge -> edge.isSameDirection(direction));
+    }
+
+    private Edge findEdgeByDirection(final Direction direction) {
```
public 과 private 메서드의 순서를 전체적으로 정렬해주시면 좋을 것 같아요 :)

ljhee92 replied:
앗 그렇네요!
주로 private 메서드라고 해서 클래스 맨 아래에 두지 않고 호출 순서대로 메서드 위치를 정렬해두는 것이 컨벤션으로 알고 있는데요.
public 메서드인 ``findNextNodeByDirection``에서 private 메서드인 ``findEdgeByDirection``를 호출하고 있음에도 불구하고 private 메서드의 순서가 더 먼저 위치하고 있었네요!
반영해두었습니다 😄

[refactor: 메서드 순서 수정](https://github.com/ljhee92/java-janggi/commit/4901175d0874d2b0dc9e0f6c92ecacff71180cba)

EastHShin commented on src/main/java/domain/board/Board.java:
```diff
@@ -0,0 +1,84 @@
+
+    public void movePiece(final Node sourceNode, final Node destinationNode, final Board board) {
+        Piece sourcePiece = findPieceByNode(sourceNode);
+        if (!sourcePiece.canMove(sourceNode, destinationNode, board)) {
+            throw new IllegalArgumentException(sourceNode + " -> " + destinationNode + " [ERROR] 이동할 수 없는 경로입니다.");
+        }
```
이동 검증시 예외에 대한 책임도 Piece에 두면 어떨까요? canMove 를 호출하는게 아닌, validate 라는 메서드를 호출하게 될 수 있겠네요

ljhee92 replied:
``Board``는 ``Piece``를 이동하는 것 자체에 대한 책임만 가지고 있고,
이동할 수 있는 경로에 대한 검증은 ``Piece`` 스스로 하는 것이 캡슐화가 강화되고 응집도가 높다고 생각되네요!
좋은 피드백 감사합니다! 😊

[refactor: 이동 경로를 검증하는 로직을 Piece 내부로 이동](https://github.com/ljhee92/java-janggi/commit/b57e1c6d78c980337173f6dd391018eec9d0d64a)

EastHShin commented on src/test/java/domain/piece/ByeongTest.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    @Test
+    void 병은_빈칸이_있는_위치로_갈_수_있다() {
```
`초나라_병은_왼쪽_빈칸으로_갈_수_있다` 테스트와 완전히 같은 테스트 로직이기도 하고, 이 테스트는 x쪽 빈칸으로 갈 수 있다 테스트들에서 포함이 이미 되는 테스트인 것 같아요

ljhee92 replied:
앗 그렇네요! 중복되는 테스트는 없어도 된다고 생각하여 바로 위 코멘트를 반영([refactor: 이동 경로를 검증하는 로직을 Piece 내부로 이동](https://github.com/ljhee92/java-janggi/commit/b57e1c6d78c980337173f6dd391018eec9d0d64a))하면서 테스트 수정 시에 제거해두었습니다👍🏻

EastHShin commented on src/main/java/domain/piece/Cha.java:
```diff
@@ -0,0 +1,63 @@
+            candidates.add(currentNode);
+        }
+
+        while (true) {
+            if (!currentNode.hasEdgeByDirection(direction)) {
+                break;
+            }
```
```suggestion
        while (currentNode.hasEdgeByDirection(direction)) {
```
while 문의 조건을 더 명확하게 표현해보면 좋을 것 같네요

ljhee92 replied:
재귀호출을 while로 바꾸면서 체크하지 못한 부분 중 하나네요🥲
while문의 조건을 명확히 해주는 것은 무한루프를 방지할 수 있어서 좋다고 생각합니다!

[refactor: while문 조건 명시](https://github.com/ljhee92/java-janggi/commit/b9014dc928987f0a6e2c8f58b684487aab76e83f)

EastHShin reviewed (approved):
안녕하세요 듀이~

피드백 반영한 부분 확인했습니다.
미션 진행 잘 해주셔서 이번 단계는 머지해도 될 것 같습니다.
몇가지 코멘트 남겼으니 확인해주시고, 다음 단계에서 뵙겠습니다 :)
